Tech Debt [PM-55] - Refactor all Identity API methods and misc auth owned api methods out of api.service monolith

JaredSnider-Bitwarden opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [ ] Bug fix
- [ ] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

<!--Describe what the purpose of this PR is. For example: what bug you're fixing or what new feature you're adding-->
[PM-55](https://bitwarden.atlassian.net/browse/PM-55)
Extract all Identity API calls (now owned by the Auth Team) into separate API Services and place them into the appropriate auth folders while refactoring code / names where appropriate to improve clarity. 

## Code changes

<!--Explain the changes you've made to each file or major component. This should help the reviewer understand your changes-->
<!--Also refer to any related changes or PRs in other repositories-->

New `api-helper.service` Service:
- **libs/common/src/services/api-helper.service.implementation.ts:** To avoid circular dependencies, I had to extract the HTTP request url creation, HTTP request creation, HTTP request execution, HTTP response handling, and misc HTTP utilities into an `api-helper.service` that the `api.service` delegates responsibilities to. 
    - The `api-helper.service` cannot leverage the new `token-api.service` to attach an authorization header to outgoing requests otherwise a circular dependency is created.  Thus, the `api.service - send(...) `method is responsible for attaching the authorization header when `authed` is true, and this maintains the existing code pattern of the `send(...)` method being the standardized way of executing an HTTP request in our application 

New Identity API (Auth) Services:
 - **libs/common/src/auth/services/token-api.service.implementation.ts:**  An API service for managing identity token exchanges (authentication) and access token refresh via refresh token / client credentials.  Sends requests against the `connect/token` endpoint (managed by IdentityServer) in the Identity API
     -  The `token-api.service` leverages the `api-helper.service` directly as it only sends "unauthed" requests to the Identity API
 - **libs/common/src/auth/services/account-api.service.implementation.ts:**  An API service for managing account information. Sends requests to the Accounts controller in the Identity API **and** to the Accounts controller in the API server project. 
 - **libs/common/src/auth/services/sso-api.service.implementation.ts:**  An API service for pre validating the SSO process. Sends requests to the SSO controller in the Identity API.

New Auth API Services:
 - **libs/common/src/auth/services/key-connector-api.service.implementation.ts:**  An API service for all key connector methods. 
 -  **libs/common/src/auth/services/auth-request-api.service.implementation.ts:**  An API service for all auth request  methods (passwordless calls).  
 
Naming changes to existing Auth Services:
- I renamed all the `userVerification` prefixed services to match the proper naming convention of `user-verification` and updated all references to the renamed services. 

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)


[PM-55]: https://bitwarden.atlassian.net/browse/PM-55?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jlf0dev commented on libs/common/src/auth/services/token-api.service.implementation.ts:
```diff
@@ -0,0 +1,174 @@
+
+  private async refreshAccessTokenViaRefreshToken(): Promise<void> {
+    const refreshToken = await this.tokenService.getRefreshToken();
+    if (refreshToken == null || refreshToken === "") {
```
Suggestion (non-blocking): There was a handy `Utils.isNullOrWhitespace` further up that you could replace this with. The error could also use some context as well.

jlf0dev commented on libs/common/src/auth/services/token-api.service.implementation.ts:
```diff
@@ -0,0 +1,174 @@
+    throw new Error("Cannot refresh token, no refresh token or api keys are stored");
+  }
+
+  private async refreshAccessTokenViaRefreshToken(): Promise<void> {
```
Suggestion: Could we pass in the refresh token like we do with client creds below with `refreshAccessTokenViaClientCredentials`? Just for consistency and reducing side effects sake.

jlf0dev commented on libs/common/src/auth/services/sso-api.service.implementation.ts:
```diff
@@ -0,0 +1,24 @@
+import { SsoApiService as SsoApiServiceAbstraction } from "../abstractions/sso-api.service.abstraction";
+import { SsoPreValidateResponse } from "../models/response/sso-pre-validate.response";
+
+export class SsoApiServiceImplementation implements SsoApiServiceAbstraction {
```
Nice! Forgot this might need a place to live as well!

jlf0dev commented on libs/common/src/auth/services/sso-api.service.implementation.ts:
```diff
@@ -0,0 +1,24 @@
+    if (identifier == null || identifier === "") {
+      throw new Error("Organization Identifier was not provided.");
+    }
+    const responseBody = await this.apiService.send(
```
Question: This is using the `apiService.send` method, but its not authed. Is there a reason why it can't just use the `apiHelperService`?

JaredSnider-Bitwarden replied:
Yes, if I use the `apiHelperService`, there is no `sendUnauthed(...)` method currently on that class (which I chose to do to avoid code duplication between a `sendUnauthed(...)` method and the existing `send(...)` method).

It was just easier to inject the existing `apiService` and leverage its `send(...)` method. 

This leaves the `token-api.service` and the `api.service` as the only two services which need the `api-helper.service`, and all other API services can leverage `api.service send(...)` as intended.

JaredSnider-Bitwarden replied:
In the future, if we re-wrote the `tokenApiService` to not handle its own errors and response parsing like a normal API service, we could potentially have use for a `sendUnauthed(...)` method on the `api-helper.service` and it could be used here also.

jlf0dev reviewed (commented):
The build is still failing so I wasn't sure if you're done yet, but I wanted to give a first pass with some suggestions. Overall it's looking great and I couldn't find any issues yet 👍

JaredSnider-Bitwarden commented:
> The build is still failing so I wasn't sure if you're done yet, but I wanted to give a first pass with some suggestions. Overall it's looking great and I couldn't find any issues yet 👍

Thank you! 

I believe the build is failing due to the test issues in master at the moment ([context](https://bitwarden.slack.com/archives/GRN5AAZ6Y/p1680276878578319)). 

I'll review your feedback shortly!

JaredSnider-Bitwarden commented on libs/common/src/auth/services/token-api.service.implementation.ts:
```diff
@@ -0,0 +1,174 @@
+        return new IdentityTokenResponse(responseJson);
+      } else if (
+        response.status === 400 &&
+        responseJson.TwoFactorProviders2 &&
```
Created a Jira Tech Debt item [PM-1696](https://bitwarden.atlassian.net/browse/PM-1696) and added a TODO referencing it as well.

JaredSnider-Bitwarden commented on libs/common/src/auth/services/token-api.service.implementation.ts:
```diff
@@ -0,0 +1,174 @@
+    try {
+      await this._refreshAccessToken();
+    } catch (e) {
+      return Promise.reject(null);
```
@jlf0dev , Yeah, the existing implementation did that, and I initially left it that way to avoid breaking anything. 

Several times, I was tempted to replace `refreshAccessToken()` with the current `_refreshAccessToken()` as I didn't see the benefit of suppressing errors, but I kept thinking maybe this was done for a reason b/c _maybe_ the access token failing to be refreshed wasn't a show stopper in some circumstance but having an error bubble up was. 

Not sure what we want to do here. 

I went through all the places `refreshAccessToken`() was called, and I found something else: It is sometimes called before full syncs... even though the `fullSync` method in the `sync.service` actually calls it as well before executing the `getSync`. So, that's fun.

jlf0dev replied:
Yeah I don't really see any benefits either but changing it might be catastrophic for some function. We can take another look later on but I don't want to expand the scope further

JaredSnider-Bitwarden replied:
@jlf0dev created a tech debt item and a TODO in code for future investigation.

JaredSnider-Bitwarden commented on libs/common/src/auth/services/token-api.service.implementation.ts:
```diff
@@ -0,0 +1,174 @@
+      return this.refreshAccessTokenViaClientCredentials(clientId, clientSecret);
+    }
+
+    throw new Error("Cannot refresh token, no refresh token or api keys are stored");
```
Good catch. I'll replace the thrown errors with rejected promises so we can pass the rejection down the promise chain to be handled in one block.

jlf0dev commented on apps/web/src/app/settings/premium.component.ts:
```diff
@@ -116,7 +118,7 @@ export class PremiumComponent implements OnInit {
 
   async finalizePremium() {
-    await this.apiService.refreshIdentityToken();
+    await this.tokenApiService.refreshAccessToken();
```
I'm guessing this is one of those instances we don't really need to refresh the access token right?

jlf0dev replied:
Yep, it looks like it's refreshing the access token anytime we actually have a sync go through, so i'm not sure we need to do it initially where we have `forceSync` set

JaredSnider-Bitwarden replied:
I'll surgically refactor that to remove the calls before force syncs.

jlf0dev reviewed (commented):
Tiny question

jlf0dev reviewed (approved):
Outstanding work! Can you make a note to QA to test around the dropped token refreshes?

MGibson1 commented on libs/common/src/abstractions/api-helper.service.abstraction.ts:
```diff
@@ -0,0 +1,47 @@
+  // to avoid circular service dependencies
+  createRequest: (
+    method: "GET" | "POST" | "PUT" | "DELETE",
+    requestUrl: string,
```
Just a thought, since this interface seems relatively complicated to me.

This could be simplified to 
```typescript
requestUrl: { path: string, apiUrl?: string }
```

and remove the `buildRequestUrl` method entirely.

MGibson1 reviewed (commented):
Are you planning on key connector-related methods in a follow up?

Other than that a few suggestions, but nothing super big, I think

JaredSnider-Bitwarden commented on libs/common/src/services/api-helper.service.implementation.ts:
```diff
@@ -0,0 +1,235 @@
+      this.device === DeviceType.VivaldiBrowser;
+  }
+
+  //#region Http Request Creation
```
> We've elected to not use regions...

Is this a documented code style? I searched contributing docs / confluence for a code style doc or similar that referenced regions, but I wasn't able to find anything. 

> ...due to feelings that they hide classes committing to too many responsibilities

Hmm.. That is a valid concern. I do see how regions _could_ obfuscate the true size of a class to the point where a dev might not realize it needs to be decomposed or a better solution re-architected.  

However, I have made heavy use of them in the past and on other PRs here for purely organizing / grouping methods within classes -- even for small classes. It's a really nice QOL feature for me to help with fast navigation + being able to order the groups in a logical manner. Further, VsCode doesn't collapse regions in TypeScript by default which alleviate some of those concerns regarding code size visibility. 

If I need to remove them to meet the guidelines of the engineering org, I will... but man, I really like their utility.

JaredSnider-Bitwarden commented on libs/common/src/auth/abstractions/token.service.ts:
Yeah, both @jlf0dev and I discussed how, when we have multiple token types (id, refresh, and access) floating around in the auth space that naming something just "token" was not ideal in terms of clarity / explicitness.  

I attempted to improve that w/ the renames, and I specifically wrestled w/ whether nor not to add `...FromAccessToken()` to the data retrieval methods. I consulted both @differsthecat && @jlf0dev, and both thought the suffix was good for additional clarity of where the data was coming from. 

Thank you for calling this out for discussion as I can sometimes err on excess verbosity.

JaredSnider-Bitwarden commented on libs/common/src/auth/services/accounts-api.service.implementation.ts:
Shoot. Good callout. 

The `account-api.service` actually goes against the API service `AccountsController` (I initially assumed that, based on the name, it went against an `AccountController`). 

The new `accounts-api.service.implementation.ts` goes against the Identity API `AccountsController`. 

However, both deal with the Account domain of data so, if we are ok with a single client side API service pointing at  different controllers in different APIs, then I will combine the two. 

From looking at the Identity API `AccountsController`, its methods were moved from the API service. Do you know if we plan to migrate additional account related methods over from the API Service to the Identity Service as that would support the combination of the two services even more?

JaredSnider-Bitwarden replied:
Per discussion with @MGibson1, we are probably not going to be moving additional accounts calls to Identity API as original methods (prelogin and postregister) were moved due to DDOS concerns. However, we are ok with consolidating all accounts endpoint calls (across both server projects) into a single client side api service for now.

JaredSnider-Bitwarden commented on libs/common/src/services/api.service.ts:
```diff
@@ -1655,61 +1554,12 @@ export class ApiService implements ApiServiceAbstraction {
-
-    if (response.status === 200) {
-      const body = await response.json();
-      return new SsoPreValidateResponse(body);
-    } else {
-      const error = await this.handleError(response, false, true);
-      return Promise.reject(error);
-    }
+    return this.apiHelperService.nativeFetch(request);
```
Ok, I can do that cleanly for the `fetch(...)` method as it is only used internally w/in the `api.service.` 

However, the `nativeFetch(...)` method is used in several places which will will require replacing the api.service dependency w/ a dependency on the `api-helper.service`.  I initially considered this a similar decision to what we decided with keeping the `api.service` as the owner of the `send(...)` (in order to avoid replacing all uses of the send method), but now I realize that it was primarily so that we could avoid the circular dependency issue w/ the `tokenApiService.getActiveAccessToken()` call. 

The `nativeFetch(...)` has no circular dependency issues so I'll get that converted over shortly. Thank you for pointing this out!

JaredSnider-Bitwarden commented on libs/common/src/auth/services/token-api.service.implementation.ts:
```diff
@@ -0,0 +1,179 @@
+    }
+  }
+
+  private async _refreshAccessToken(): Promise<void> {
```
I had to prefix it with "_" to avoid a compilation error regarding duplicate function implementation. TypeScript doesn't allow the same method name (even if one is public and another is private). 
![image](https://user-images.githubusercontent.com/116684653/230231794-963aa3f5-5a72-4709-adb6-fd1ec894aba5.png)

JaredSnider-Bitwarden commented:
@MGibson1 
> Are you planning on key connector-related methods in a follow up?

Yes, great callout. I've added a [tech debt item ](https://bitwarden.atlassian.net/browse/PM-1783)to track this in our backlog and to fix at a later date.

eliykat reviewed (commented):
I've glanced over this from AC Team's perspective, it looks like minor changes only, so our files are good.

Others have reviewed in more detail so I haven't gone any further, but overall I appreciate the effort to break up apiService.

JaredSnider-Bitwarden commented on libs/common/src/auth/login-strategies/login.strategy.ts:
ApiService has been removed from the login strategies to meet the original prompt of the PM-55 task. Thank you again for the reminder for this!

JaredSnider-Bitwarden commented on libs/common/src/auth/services/auth.service.ts:
API service dependency completely removed from the auth service.

JaredSnider-Bitwarden commented:
PR is ready for 2nd review.

jlf0dev commented:
> Please split the PR into two or more
@Hinton while the size of this PR has become considerable, and there are two separate concerns here, I'm not sure there's enough value in the time spent breaking it down into two separate PRs with another round of reviews. There were 4 other reviews on the code up to the point where @JaredSnider-Bitwarden started breaking out the api services, which seems to be partly why it grew so much. 

> I wonder if there aren't better approaches to do this
Introducing the circular dependency again seems like a step backward to me at this point. We can definitely add another tech debt task as well though if there are others that agree. I would be interested in hearing @JaredSnider-Bitwarden's perspective on that idea as well.

JaredSnider-Bitwarden commented:
@Hinton, I am sorry for the larger size of this PR. Here's the context for the history of the work:

The initial scope was to "Move log in strategy API methods to an owned file." However, early on in the work (done initially in another, now deleted branch) I ran into the circular dependency issues early w/ the creation of the `token-api.service`, and it necessitated the creation of the `api-helper-service` as a solution (note: the design was settled on after multiple meetings with @differsthecat , @jlf0dev , and @MGibson1). Once we understood the issue at hand fully, I restarted the work in this branch to create a more easily reviewable commit history. 

Then once past the initial review stages, @MGibson1 correctly pointed out that I'd not quite accomplished the original prompt of extracting all the login strategy API methods into their own files. So, I finished extracting the auth methods to resolve the PR feedback. 

As part of this response, I've updated the PR description to try and provide additional clarity regarding the core service changes / extractions.  Please let me know if I can clarify any further aspects of the changes!

***

> injecting the service in the method call...

I feel strongly that separating out the concerns into separate, single responsibility services is a more maintainable, best practice solution vs re-introducing the circular dependency and solving via passing a service as a parameter and sidestepping Angular's dependency injection (which creates tight coupling between services which is less maintainable / testable imho). 

@MGibson1 and I also discussed ideas for what the endgame architecture for the `api-helper.service` might look like. In the future once the `api.service` has been fully decomposed into separate api services, we could revisit the new architecture to see if we could improve the multi service design to solve the authed vs unauthed request issue more cleanly without re-introducing a circular dependency at that time.

Hinton reviewed (changes requested):
I'll get this checked out locally to review the ApiService changes properly sometime next week since there are to many files for the GitHub UI to function. But the changes as is won't work on node since we're not overriding the fetch method.

JaredSnider-Bitwarden commented:
@Hinton 
> there are to many files for the GitHub UI to function.
Thank you for bringing that to my attention; I did not realize that. I'm breaking apart the PR to mitigate that and to increase ease of review.  

I'm also investigating a solution to the `NodeApiService` extending the `ApiService` to override the `nativeFetch(...)` method which no longer exists on the `ApiService` as it has been moved to the `ApiHelperService`/

JaredSnider-Bitwarden commented:
Closing this as we are going to break it up into two consecutive PRs. First PR is [here](https://github.com/bitwarden/clients/pull/5218).
